examples/custom_entry_buffer/entry_buffer/imp.rs

```diff
@@ -0,0 +1,128 @@
+use std::rc::Rc;
+
+#[derive(Debug)]
+pub struct CustomEntryBuffer {
+    text: Rc<RefCell<String>>,
+    length: Cell<u16>,
+    max_length: Cell<u16>,
+}
+
+impl Default for CustomEntryBuffer {
+    fn default() -> Self {
+        Self {
+            text: Rc::new(RefCell::new(String::new())),
+            length: Cell::new(0),
+            max_length: Cell::new(0),
+        }
+    }
+}
+
+#[glib::object_subclass]
+impl ObjectSubclass for CustomEntryBuffer {
+    const NAME: &'static str = "CustomEntryBuffer";
+    type Type = super::CustomEntryBuffer;
+    type ParentType = gtk::EntryBuffer;
+}
+
+impl ObjectImpl for CustomEntryBuffer {
+    fn properties() -> &'static [ParamSpec] {
+        static PROPERTIES: Lazy<Vec<ParamSpec>> = Lazy::new(|| {
+            vec![
+                ParamSpec::new_uint(
+                    "length",
+                    "length",
+                    "Length",
+                    0,
+                    u16::MAX as u32,
+                    0,
+                    ParamFlags::READABLE,
+                ),
+                ParamSpec::new_uint(
+                    "max-length",
+                    "Maximum length",
+                    "Maximum number of characters for this entry",
+                    0,
+                    u16::MAX as u32,
+                    0,
+                    ParamFlags::READWRITE | ParamFlags::EXPLICIT_NOTIFY,
+                ),
+                ParamSpec::new_string(
+                    "text",
+                    "Text",
+                    "The contents of the buffer",
+                    Some(""),
+                    ParamFlags::READWRITE,
+                ),
+            ]
+        });
+        PROPERTIES.as_ref()
+    }
+
+    fn property(&self, _obj: &Self::Type, _id: usize, pspec: &ParamSpec) -> Value {
+        match pspec.name() {
+            "length" => (self.length.get() as u32).to_value(),
+            "max-length" => (self.max_length.get() as u32).to_value(),
+            "text" => self.text.borrow().to_value(),
+            _ => unreachable!(),
+        }
+    }
+
+    fn set_property(&self, _obj: &Self::Type, _id: usize, value: &Value, pspec: &ParamSpec) {
+        match pspec.name() {
+            "length" => {
+                self.length.set(value.get::<u32>().unwrap() as u16);
+            }
+            "max-length" => {
+                self.max_length.set(value.get::<u32>().unwrap() as u16);
+            }
+            "text" => {
+                self.text.replace(value.get().unwrap());
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl EntryBufferImpl for CustomEntryBuffer {
+    fn text(&self, _entry_buffer: &Self::Type) -> glib::GString {
+        self.text.borrow().clone().into()
+    }
+
+    fn length(&self, _entry_buffer: &Self::Type) -> u16 {
+        self.text.borrow().chars().count() as u16
+    }
+
+    fn insert_text(&self, entry_buffer: &Self::Type, _position: u16, chars: &str) -> u16 {
```

Comment:
Both the current insert_text & delete_text methods are wrong due to their complexity on figuring out the bytes position back from the chars position where to insert/delete the string. I was wondering if there might some kind of API/external crate that helps for such situations?

Reply:
The `std` API is not enough? How do actual implementors of this in C handle it?